src/jcpsantiago/bulgogi.clj

```diff
@@ -31,7 +48,8 @@
   "
   [req -ns]
   (let [{:keys [input-data features]} req
-        fns (resolved-features features -ns)
+        fns (memoized-features features -ns)
+        co-effects (memoized-co-effects fns)
         fn-ks (map keyword features)]
-    (->> (transformed input-data fns)
+    (->> (transformed (enriched input-data co-effects) fns)
```

Comment:
we are limited to one dependency level, right? e.g. say `a` needs `b` which needs `c`, `enriched` will try to apply `b` and `c` to the data, but it's possible they are not executed in the correct order (if I'm seeing this correctly). There's no problem for simple relationships `a<-b`

Reply:
Yes, I've built it to be one level right now. Recursing on the dependencies would be easy to implement though.
Currently coeffect functions are treated differently than feature functions, they don't get to have rely on coeffects themselves. I wasn't sure how complicated the system is allowed to be, so I started with the minimum feature set.
I'd be happy to extend it so that coeffect functions can rely on other coeffects, the basic version is pretty trivial to implement. There are some design decisions about circular dependencies and idempotence though.

Reply:
yeah I think this is a great start, I'm also not sure how complex this needs to be in the real world.. from my experience at Billie this is already enough